lib/http2/hpack.c

```diff
@@ -378,8 +385,11 @@ int h2o_hpack_decode_header(h2o_mem_pool_t *pool, void *_hpack_header_table, h2o
 
     /* determine the value (if necessary) */
     if (!value_is_indexed) {
-        if ((value = decode_string(pool, src, src_end, 0, err_desc)) == NULL)
+        if ((value = decode_string(pool, src, src_end, 0, err_desc)) == NULL) {
+            /* this may leak `name` but as long as it is tied to the pool, it will be soon freed at pool destruction,
+             * so it is harmless to ignore the leak */
```

Comment:
I would not call it a leak, and rather avoid adding this type of comment.

This is the consistent way of we use memory pool. I can understand that the confusion some might initially have, but the point is that if you do not understand the API, that confusion happens everywhere. And we cannot add this comment everywhere.

Reply:
Right... the above comment was based on my incorrect understanding on how the pool works.

How about simply adding `assert(pool != NULL);` here? If `pool` is not used (I'm not sure if we ever experience such a usage though), this may cause actual leak.

Reply:
First, let me point out that `pool` never becomes NULL.

If `pool` can become NULL relates to what the contract of the public function that call `decode_string`. As we know, the only caller is `h2o_hpack_decode_header`. By looking at `h2o_hpack_decode_header`, we see  functions like `h2o_mem_link_shared` being invoked without checking if `pool` is NULL. To paraphrase, there is an assumption that `pool` would never be NULL.

Actually, we can assume that functions designed specifically for HTTP request processing would have access to a memory pool. This is because reducing the cost of memory allocation is essential for high-performance HTTP request processing.

I think that's the reason why we did not even think of clarifying the condition anywhere. As a side note, I think it would be a good idea to be principle on clarifying optional pointers in the doc-comments, like we do in [string_.h](https://github.com/h2o/h2o/blob/master/include/h2o/string_.h#L48-L55).

Regarding if we should be adding `assert(pool != NULL)`, I do not think we need to add that to avoid a leak considering that such condition might cause a crash elsewhere (as stated above). I also do not think that we need to add an assert in `h2o_hpacke_decode_header`, because we do not have a principle of testing API contracts using assertions.

Reply:
I see. I was too much focusing on the memory pool implementation and having a wrong impression that `pool` could potentially become `NULL` regardless of the context (caller). I'll remove code and comments assuming pool could potentially be NULL.